doc: drop support for VS2015

seishun opened the pull request:
V8 master no longer builds with VS2015: https://github.com/nodejs/node-v8/issues/19#issuecomment-342440529 There are two main issues:

* Internal compiler error. Might get fixed by Microsoft, but chances are slim since VS2015 is no longer actively supported. There might be a workaround, but the V8 team is probably not eager to spend time tinkering to please an outdated compiler.
* Extended `constexpr` in `bits.h` (introduced in https://github.com/v8/v8/commit/51f4d2e9e3459a5f3f92f24183d52f8d84137596). Supported since VS2017. There is no trivial way to work around it while keeping the function `constexpr`.

Visual Studio Community 2017 is free and supports all the operating systems VS2015 does, so spending extra effort to support VS2015 doesn't seem justifiable.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
doc

gibfahn commented:
Previous issue/PR (to move from 2013 to 2015) was https://github.com/nodejs/node/issues/7484 and https://github.com/nodejs/node/pull/8049

In that @joaocgreis confirmed that node compiled with 2015 still worked with modules compiled with 2013 (https://github.com/nodejs/node/issues/7484#issuecomment-235659948). We should probably do something similar.

From https://github.com/nodejs/node/issues/7484#issuecomment-249307034, it sounds like node-sass is a good test-case, cc @saper 

FWIW now seems like a good time to do this, and if V8 is forcing our hand then +1 on this for Node 10.0.

seishun commented:
@gibfahn You sure about semver-major? VS 2013 was dropped in Current (https://github.com/nodejs/node/issues/7989).

gibfahn commented:
>@gibfahn You sure about semver-major? VS 2013 was dropped in Current (#7989).

Marking it defensively, if it turns out that it's still compatible with native addons built with VS 2013 then I think it's not semver-major.

refack commented:
AFAIR the decision to make this `major` was that we build the release binaries with the oldest supported toolset. If we drop support we should first start building with a newer toolset and that _could_ be a `major` change (native modules _might_ need to be recompiled).

Besides that:
1. I'm pro dropping VS2015 as soon as it's possible
2. As long as MS distributes a free variant of it's toolchain, we should change policy to official support only one toolchain (continue CI with what is possible/convenient)
3. Open bugs for V8 if the build breaks with the toolchains we use. (Refs: https://chromium-review.googlesource.com/c/v8/v8/+/753903)

/CC @nodejs/v8

refack commented:
@seishun I added two commits that remove more mentions of VS2015. Feel free to push them out.

seishun commented:
@refack The discussion in https://github.com/nodejs/node/issues/14798 is still ongoing and "Allow edits from maintainers" seems to be checked by default, so I'd avoid pushing to PR branches for now.

Also, I disagree with squashing Build Tools into "any edition" since it's not an edition of VS2017. Plus "Build Tools SKU" is a confusing name (what exactly is "SKU"?).

refack commented:
> The discussion in #14798 is still ongoing and "Allow edits from maintainers" seems to be checked by default, so I'd avoid pushing to PR branches for now.
> 
> Also, I disagree with squashing Build Tools into "any edition" since it's not an edition of VS2017. Plus "Build Tools SKU" is a confusing name (what exactly is "SKU"?).

As I wrote, I added this only as a suggestion, feel free to push it out, or take what you like. I only moved some sentences around, did not add anything new, and I don't feel strongly about any of the phrasing.
The sentence with "SKU" was in the "Windows prerequisites" section. I like it because it short, and because for node's needs there is no difference between the "Visual Studio IDE" and "Build Tools for Visual Studio 2017".

AFAIK SKU is is a term used like catalog-number, that is a label given to a specific version of a product including all its parts. Microsoft uses it as the full name of each variation of it's products (for example Product = `Visual Studio`, Version = `2017`, Edition = `Community`, SKU = `Visual Studio 2017 Community Edition`).

gibfahn commented:
>@refack The discussion in #14798 is still ongoing and "Allow edits from maintainers" seems to be checked by default, so I'd avoid pushing to PR branches for now.

FWIW it's become pretty common to push suggestions to branches (obviously not force pushing), I should probably update that issue to reflect that.

joaocgreis commented:
There was already some discussion in https://github.com/nodejs/node/issues/13052#issuecomment-331008660 (but now we're actually dropping VS2015).

Do we need to land a version of V8 that doesn't work with VS2015 in v9.x? If not, it would be better to move as semver major in v10.0.0. This is because we're hopeful that there are no issues with ABI compatibility for native modules compiled with VS2015, but we can't be completely sure.

There are still some steps that need to happen before this PR can land:
- remove VS2015 from `vcbuild.bat` (would be nice to have as part of this PR)
- add a release machine with VS2017 (I'll move forward with this one, so that we can start building nightlies soon)
- test native modules (as mentioned by @gibfahn above) to ensure that users can still use VS2015 to build native modules, as it's too early to drop that. This is essentially compiling node in a machine with VS2017 and then moving the whole folder to a machine with VS2015 (so that `--nodedir` can be used), running CitGM and taking a close look at the results. Adding support for this to the CitGM CI job would also be great, but that would require cross compiling, perhaps we'll have to leave that to a next stage. @seishun as you said in IRC, want to give this a try?
- adapt the current test job in CI, but that's easy after all of the above.

seishun commented:
>remove VS2015 from vcbuild.bat (would be nice to have as part of this PR)

I was thinking about doing it in a separate PR to avoid mixing discussions.

>test native modules (as mentioned by @gibfahn above) to ensure that users can still use VS2015 to build native modules, as it's too early to drop that.

This is necessary to make this non-major, right?

>This is essentially compiling node in a machine with VS2017 and then moving the whole folder to a machine with VS2015 (so that --nodedir can be used), running CitGM and taking a close look at the results.

Why does it need to be a separate machine? What if I just build node with VS2017, then make sure CitGM runs with VS2015?

>@seishun as you said in IRC, want to give this a try?

Sure, but ideally I think it should be done by someone with more experience with these things.

seishun commented:
I have no idea how to run `citgm` only on native addons, so I just ran `citgm node-sass --nodedir=<path to folder>`. I got "canary is dead", followed by a few hundred warnings and then

```
warn:                        | npm ERR! Test failed.  See above for more details.                                       
info:    done                | The smoke test has passed.
```

Not sure if it passed or failed. Also, I got the same result when I ran it on the machine with VS2017. Plus it's unclear if it even built the binary since `npm install node-sass` downloads it.

Since `citgm` isn't exactly intuitive, I just ran `node-gyp rebuild --nodedir=<folder>` and ran the tests manually. Got two `EEXIST` failures that didn't look related to ABI.

Anyway, I don't want to spend more time on this since [this blog post](https://blogs.msdn.microsoft.com/vcblog/2017/03/07/binary-compatibility-and-pain-free-upgrade-why-moving-to-visual-studio-2017-is-almost-too-easy/) and [these](https://stackoverflow.com/a/40896092/550110) [two](https://stackoverflow.com/a/44314010/550110) answers claim that VS2017 is binary compatible with VS2015, which should be sufficient. I'm removing the `semver-major` label.

refack commented:
> I have no idea how to run citgm only on native addons, so I just ran citgm node-sass --nodedir=<path to folder>. I got "canary is dead", followed by a few hundred warnings and then

The main "power" of CitGM is it's doing batch tests. Basicly what it does is downloads a package, runs npm install, then npm test. You can do that manually and the output should be more readable.

refack commented:
BTW: removing VS2015 support from `vcbuild.bat` is a semver-major change, so I agree with @seishun that it should be done in a separate PR.

seishun commented:
>removing VS2015 support from vcbuild.bat is a semver-major change

https://github.com/nodejs/node/pull/8067 wasn't semver-major, and it doesn't make sense to keep VS2015 "support" in vcbuild in 9.x if it won't actually compile.

richardlau commented:
> I have no idea how to run citgm only on native addons,

This should do the trick:
```citgm-all --includeTags native```

joaocgreis commented:
> [this blog post](https://blogs.msdn.microsoft.com/vcblog/2017/03/07/binary-compatibility-and-pain-free-upgrade-why-moving-to-visual-studio-2017-is-almost-too-easy/) and [these](https://stackoverflow.com/a/40896092/550110) [two](https://stackoverflow.com/a/44314010/550110) answers claim that VS2017 is binary compatible with VS2015

That's helpful. So, v140 and v141 are guaranteed to be binary compatible, which should make this transition much less risky. The [second answer](https://stackoverflow.com/a/44314010/550110) leaves it quite clear that we'll have the issue again in v150, hopefully by then N-API will help mitigate this.

I still think we shouldn't rush this. Let me get the release machine ready, we'll start building nightlies of master (v10) with it and see if any issue surfaces. Is this needed for v9?

>>remove VS2015 from vcbuild.bat (would be nice to have as part of this PR)
>
>I was thinking about doing it in a separate PR to avoid mixing discussions.

How is it different discussion? I would do it in the same commit, because one does not make sense without the other. If one is semver major, how can the other not be?

saper commented:
N-API is just a first step. C++ runtime support will be needed even if pure C interfaces will be used and that's why it is still important to keep binary compatibility between components.

refack commented:
If we assume compiled addons will be compatible, as I see it there are two issues
1. Dropping "official" support, recommending VS2017, but keep testing on VS2015 (IMHO this is semver-minor similar to documentation only deprecation)
2. Breaking VS2015 support, and stop testing on VS2015 (IMHO semver-major, similar to removal of a feature)

I'm +1 on this PR as is, since it's only addresses issue (1).
Issue (2) should be postponed for as long as we can.

seishun commented:
@joaocgreis 

>I still think we shouldn't rush this. Let me get the release machine ready, we'll start building nightlies of master (v10) with it and see if any issue surfaces.

We run CI on VS2017, what kind of issues do you expect other than build infra issues?

>Is this needed for v9?

If V8 is going to be upgraded in 9.x, then yes.

>How is it different discussion?

There _might_ be some technical discussion regarding the `vcbuild.bat` file, I'd prefer to separate it from this. But if anyone else feels strongly about it, I don't mind combining both.

@refack

>IMHO semver-major, similar to removal of a feature

I refer to @bnoordhuis's comment [above](https://github.com/nodejs/node/pull/16868#pullrequestreview-75052931) which I fully agree with.

refack commented:
>> IMHO semver-major, similar to removal of a feature
>
> I refer to @bnoordhuis's comment above which I fully agree with.

I not sure either way, I would advise for cautions, but I'm not blocking either way.

As for issues, there is at least one bug that reproduces only in VS2017 build binaries https://github.com/nodejs/node/issues/15558#issuecomment-333278380

seishun commented:
I've updated the commit to include necessary changes in the other sections. I've also removed the space in "VS2015" because it looks nicer.

I wasn't sure how to update the "Visual Studio 2015 or Visual C++ Build Tools 2015 or newer" line. It's not clear what the actual name of the Build Tools product is - it's "Build Tools for Visual Studio 2017" in the downloads page, and "Visual Studio Build Tools 2017" in the installer, so I just went with a short version. I've also dropped "and newer" since there's nothing newer than 2017.

@nodejs/platform-windows I'm going to land this in 24 hours if there are no objections.

refack commented:
> I've also dropped "and newer" since there's nothing newer than 2017.

👍 that's prudent as build might not work out of the box with the next version.

hashseed commented:
Having only skipped through this thread, is there a consensus on requiring VS2017 for Node.js at this point? We are currently adding Windows and Mac CI bots on our end at V8, and it would be awesome if we didn't have to set up an additional one for VS2017. Chromium no longer requires VS2015, for reference.

benjamingr commented:
@hashseed apparently V8 contains code that doesn't compile in VS2015 to begin with - so even if we wanted to (short of Node.js forking V8 or PRing against that file) - we can't support VS2015.

Given VS2017 is free and supports exactly the same platforms as VS2015 the consensus in this thread is to support only VS2017 moving forward. This is beneficial for Node.js anyway since it is one less thing to maintain.

hashseed commented:
I get that part. I'm trying to get a feeling of whether that's a regression we can live with, by dropping VS2015 support in Node.js, or something we need to fix in V8 until Node.js drops support.

The current situation is unfortunate in that V8 went ahead and stopped testing for VS2015, and we broke Node.js' expectations. The question is where we go from here.

benjamingr commented:
> From your response I gather that we can just drop support for VS2015 like we already have.

Yeah, I agree it's unfortunate but given there are no platforms VS2015 runs that VS2017 doesn't run on I don't think it's _that_ unfortunate either :)

seishun commented:
Landed in https://github.com/nodejs/node/commit/c5a49e148d3293eb9e8c17a15cb8c876977f76af.

refack commented:
> I get that part. I'm trying to get a feeling of whether that's a regression we can live with, by dropping VS2015 support in Node.js, or something we need to fix in V8 until Node.js drops support.

@hashseed it depends on whether you CI 6.2 and 6.3, as we are dropping official support for VS2017, which is certain for node10, but it is still left to be decided if we can start building node8 and node9 with VS2017. So it only the tip of V8 is broken, it's possible to live with that, but it might limit our options for upgrading V8 for the active version lines (TDB).

As a post mortem, I was wondering if we could get a heads up when chromium/V8 bumps requirements, so that we can follow suit as soon as possible.

hashseed commented:
Since we are only trying to make sure updating V8 does not break Node.js, we only test updating V8 in a reasonably recent version of Node.js (which we sync with upstream every 2-3 months). That seems to have worked well so far aside from this Windows build issue.

I'll keep in mind next time V8 changes build requirements.

joaocgreis commented:
> We run CI on VS2017, what kind of issues do you expect other than build infra issues?

CI runs only node's addon tests. CitGM is only run for releases, but even that is no guarantee that nothing in the wild will break. The [issue in node-sass](https://github.com/sass/node-sass/issues/1283) with VS2015 shows that compatibility issues can cause a lot of pain and be very hard to debug. Even with the guarantee of binary compatibility, I'd rather err on the safe side. Furthermore, there's the possibility of issues with VS2017 like the one @refack pointed.

Doing a documentation only removal still doesn't make complete sense to me, as this is something that we can know for sure, not a platform that we'll keep working while there are only minor issues to address. But it's good as a way to start warning, if it turns out that V8 actually needs to be updated to an incompatible version in v9.x.

@seishun please don't land PRs in a hurry when there's no need to. This PR was open only 5 days and discussion was still taking place.

@hashseed I think we're generally ok with moving node to VS2017, for compiling node core. A different issue is requiring it for users, who need to compile native modules. Even if we compile node core with VS2017, there can be cases where the V8 headers use features that are not supported by VS2015, thus modules will fail to compile with it. Ref https://github.com/nodejs/node-v8/issues/4 for an issue like this with VS2013. It would be great if this could be tested on the V8 side as well.

MylesBorins commented:
I'm marking don't land for 6.x and 8.x as I'm assuming we will not be changing the VS support for either

please LMK if this is a mistake
